File: zircon/system/core/devmgr/fshost/metrics.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "metrics.h"

#include <lib/async/cpp/task.h>

#include <type_traits>

#include <cobalt-client/cpp/metric-options.h>

namespace devmgr {
namespace {
cobalt_client::MetricOptions MakeMetricOptions(fs_metrics::Event event) {
  cobalt_client::MetricOptions options;
  options.metric_id = static_cast<std::underlying_type<fs_metrics::Event>::type>(event);
  options.event_codes = {0, 0, 0, 0, 0};
  return options;
}
}  // namespace

FsHostMetrics::FsHostMetrics(std::unique_ptr<cobalt_client::Collector> collector)
    : collector_(std::move(collector)) {
  counters_.emplace(fs_metrics::Event::kDataCorruption,
                    std::make_unique<cobalt_client::Counter>(
                        MakeMetricOptions(fs_metrics::Event::kDataCorruption), collector_.get()));
}

FsHostMetrics::~FsHostMetrics() {
  if (collector_ != nullptr) {
    collector_->Flush();
  }
}

void FsHostMetrics::LogMinfsCorruption() {
  counters_[fs_metrics::Event::kDataCorruption]->Increment();
}

void FsHostMetrics::FlushUntilSuccess(async_dispatcher_t* dispatcher) {
  if (!collector_->Flush()) {
    async::PostDelayedTask(
        dispatcher, [this, dispatcher]() { FlushUntilSuccess(dispatcher); }, zx::sec(10));
  }
}

}  // namespace devmgr

File: tools/check-licenses/lib/metrics.go
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package lib

import "fmt"

// Metrics is used for instrumentation
type Metrics struct {
	values map[string]uint
	order  []string
}

func (metrics *Metrics) Init() {
	metrics.values = make(map[string]uint)
	metrics.order = []string{
		"num_extensions_excluded",
		"num_licensed",
		"num_matched_to_project_file",
		"num_non_single_license_files",
		"num_single_license_file_match",
		"num_single_license_files",
		"num_unlicensed",
		"num_with_project_license",
	}
	for _, key := range metrics.order {
		metrics.values[key] = 0
	}
}

func (metrics *Metrics) increment(key string) {
	if _, found := metrics.values[key]; !found {
		fmt.Printf("error: metric key (%s) not found\n", key)
	}
	metrics.values[key]++
}

func (metrics *Metrics) print() {
	fmt.Println("Metrics:")
	for _, value := range metrics.order {
		fmt.Printf("%s: %d\n", value, metrics.values[value])
	}
}
